clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -4033,9 +4098,11 @@ void Sema::CheckSYCLKernelCall(FunctionDecl *KernelFunc,
 
 // For a wrapped parallel_for, copy attributes from original
 // kernel to wrapped kernel.
-void Sema::copySYCLKernelAttrs(const CXXRecordDecl *KernelObj) {
+void Sema::copySYCLKernelAttrs(const CXXRecordDecl *KernelObj,
+                               FunctionDecl *KernelCallerFunc) {
   // Get the operator() function of the wrapper.
-  CXXMethodDecl *OpParens = getOperatorParens(KernelObj);
+  CXXMethodDecl *OpParens =
+      getCallOperatorInvokedFromKernel(KernelObj, KernelCallerFunc, *this);
```

Comment:
I'm a bit concerned that `getCallOperatorInvokedFromKernel` is called so many times and each time it builds a callgraph for the whole TU. Can the callgraph be saved somewhere and re-used?
Also, can the callgraph be build not for the translation unit, but starting from the function with `sycl_kernel` attribute?

Reply:
>Also, can the callgraph be build not for the translation unit, but starting from the function with sycl_kernel attribute?

Good tip. I will try to set the call graph from the kernel root.

>I'm a bit concerned that getCallOperatorInvokedFromKernel is called so many times and each time it builds a callgraph for the whole TU. Can the callgraph be saved somewhere and re-used?

I toyed with this thought of constructing the call graph just once and passing it to `getCallOperatorInvokedFromKernel` 
However, `getCallOperatorInvokedFromKernel()` is only called 3 times, and it didn't seem too bad. WDYT?

Reply:
> However, getCallOperatorInvokedFromKernel() is only called 3 times, and it didn't seem too bad.

AFAIK, building callgraph is expensive and we shouldn't be doing that in FE at all. The less callgraphs we build in FE, the better.

Reply:
@Fznamznon I have re-written the code using `RecursiveASTVisitor`
Please let me know if this is ok.

Reply:
> Please let me know if this is ok.

Yes, looks reasonable.